### PR TITLE
Fix centos install

### PR DESCRIPTION
**Pull request by jodh-intel**

### CODE REVIEW

**Comment by chavafg** [comment]
qa-passed

**Comment by chavafg** [comment]
qa-passed

**Comment by chavafg** [comment]
qa-passed

**Comment by jodh-intel** [comment]
Hi @GabyCT - could you test this PR along with atleast 1 reviewer please?

For reference, I used a clean CentOS 7.3-1611. Specifically, I did this:

- Install using `CentOS-7-x86_64-Minimal-1611.iso`

- Perform very mimimal setup:
```
$ sudo yum -y update
$ sudo yum install -y git
$ mkdir -p $HOME/go/src/github/01org
$ cd $HOME/go/src/github/01org
$ git clone https://github.com/01org/cc-oci-runtime
$ cd cc-oci-runtime
```
- Run the install script:
```
$ script -fec ./installation/rhel-setup.sh
```
- Created a Clear Container:
```
$ sudo docker run -ti busybox sh
```

**Comment by chavafg** [comment]
worked perfectly for me using steps above. 
I used Centos7 cloud image.

**Comment by amshinde** [comment]
So, I tried this script on a clean CentOS 7.3-1611 VM. Had this running for a while.

I got this error at cc-oci-runtime compilation stage:
checking whether the C compiler works... no
configure: error: in /home/amshinde/gopath/src/github.com/01org/cc-oci-runtime:
configure: error: C compiler cannot create executables

Same kind of error for qemu-lite: 
ERROR: "/usr/local//bin/gcc" cannot build an executable (is your linker broken?)

This seems to be where gcc build failed:

make[2]: Leaving directory /home/amshinde/go/src/github/01org/cc-oci-runtime/gcc-6.2.0/host-x86_64-pc-linux-gnu/libcc1
/bin/sh: line 3: cd: x86_64-pc-linux-gnu/libmpx: No such file or directory
make[1]: *** [install-target-libmpx] Error 1
make[1]: Leaving directory /home/amshinde/go/src/github/01org/cc-oci-runtime/gcc-6.2.0


Maybe "--disable-libmpx" will help.

**Comment by chavafg** [comment]
qa-passed

**Comment by chavafg** [comment]
qa-passed

**Comment by jodh-intel** [comment]
Hi @amshinde - I've re-tested but cannot recreate the problem you saw. I'm guessing your system might be suffering from `ENOSPC`?

After careful perusal of the typescript, I found and fixed a couple of issues:

- `$deps_dir`

  This was not being created meaning the "outer" `pushd`/`popd` calls weren't working. This issue was non-fatal though.

- `yum -y update`.

  This was being called twice. The second instance was actually upgrading docker from `docker-engine-1.12.1-1.el7.centos.x86_64` to `docker-engine-17.04.0.ce-1.el7.centos`. This should have been a problem since we don't yet support docker 1.13... and yet everything appeared to work. I could add in the following to the script to pin docker to v1.12.1, but I'm now wondering if that's actually necessary given the correct behaviour I'm seeing:
  ```
  $ sudo yum -y install yum-versionlock
  $ sudo yum -y install docker-engine-1.12.1-1.el7.centos.x86_64
  $ sudo yum versionlock docker
  ```

**Comment by jodh-intel** [comment]
@devimc - you might be interested in my update above.

**Comment by chavafg** [comment]
qa-passed

**Comment by GabyCT** [comment]
@jodh-intel, I ran it using RHEL however I am having an error when you are trying to build qemu as it says configure: error in /root/gopath/src/github.com/01org/cc-oci-runtime, configure error: C compiler cannot create executables, checking wheter the C compiler works ...no

**Comment by GabyCT** [comment]
@jodh-intel it looks like something like the @amshinde error about the compiler by the way I am using the installation on baremetal

**Comment by jodh-intel** [comment]
Hi @amshinde - can you confirm that you have space in `/home/`?
Hi @GabyCT  - can you confirm that you have space in `/root/`? Also, that suggests you are running the script as root. Please can you try running it as a non-root user as the script will switch to `root` when it needs to (using `sudo(8)`).

**Comment by amshinde** [comment]
@jodh-intel. Still have space left in /home.

$ df -hT /home
Filesystem                Type  Size  Used  Avail Use%     Mounted on
/dev/mapper/cl-root  xfs   6.2G  4.5G  1.8G  73%       /

**Comment by jodh-intel** [comment]
@GabyCT / @amshinde - can you attach or send me your complete `typescript` files from running `rhel-setup.sh`?

**Comment by GabyCT** [comment]
@jodh-intel I have plenty of space as this is baremetal...I will run it as another user

**Comment by jodh-intel** [comment]
Do you still have the `typescript` files?

**Comment by amshinde** [comment]
Yes, sent it to you.

**Comment by GabyCT** [comment]
@jodh-intel sent it :)

**Comment by jodh-intel** [comment]
Hi @GabyCT - thanks for your `typescript`. Your failure seems to have been caused due to GMP not being installed which led the gcc build to fail. The reason for the install failure looks like the user you ran the test as was not able to run `sudo(8)` without a password, so `sudo` timed out and failed the install step.

Hi @amshinde - thanks too for your `typescript`. Firstly, I noticed from the log that you are not using the latest version of my `fix-centos-install` branch.

The snippet below is what is causing the failure for you I believe:

```
++ -fno-PIE -c  -DIN_GCC_FRONTEND -g -O2 -DIN_GCC     -fno-exceptions -fno-rtti -fasynchronous-unwind-tables -W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual -Wmissing-format-attribute -Woverloaded-virtual -pedantic -Wno-long-long -Wno-variadic-macros -Wno-overlength-strings   -DHAVE_CONFIG_H -I. -Ic-family -I../.././gcc -I../.././gcc/c-family -I../.././gcc/../include -I../.././gcc/../libcpp/include  -I../.././gcc/../libdecnumber -I../.././gcc/../libdecnumber/bid -I../libdecnumber -I../.././gcc/../libbacktrace   -o c-family/c-common.o -MT c-family/c-common.o -MMD -MP -MF c-family/.deps/c-common.TPo ../.././gcc/c-family/c-common.c
g++ -fno-PIE -c  -DIN_GCC_FRONTEND -g -O2 -DIN_GCC     -fno-exceptions -fno-rtti -fasynchronous-unwind-tables -W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual -Wmissing-format-attribute -Woverloaded-virtual -pedantic -Wno-long-long -Wno-variadic-macros -Wno-overlength-strings   -DHAVE_CONFIG_H -I. -Ic-family -I../.././gcc -I../.././gcc/c-family -I../.././gcc/../include -I../.././gcc/../libcpp/include  -I../.././gcc/../libdecnumber -I../.././gcc/../libdecnumber/bid -I../libdecnumber -I../.././gcc/../libbacktrace   -o c-family/c-cppbuiltin.o -MT c-family/c-cppbuiltin.o -MMD -MP -MF c-family/.deps/c-cppbuiltin.TPo ../.././gcc/c-family/c-cppbuiltin.c
make[2]: *** [s-attrtab] Killed
make[2]: *** Waiting for unfinished jobs....
```
My guess is that your VM doesn't have sufficient memory assigned to it and the OOM kicked in.

For reference, I'm testing in a qemu VM:

- installed using `CentOS-7-x86_64-Minimal-1611.iso`.
- with 2G memory.
- with 20GB disk.

**Comment by jodh-intel** [comment]
I've re-run the steps outlined in https://github.com/01org/cc-oci-runtime/pull/799#issuecomment-292164923 and have a working system that can start Clear Containers.

**Comment by jodh-intel** [comment]
It's worth pointing out that you might get a "soft error" (harmless) about `isl` at the gcc config stage like this:

```
checking for isl 0.16, 0.15, or deprecated 0.14... no
recommended isl version is 0.16 or 0.15, the minimum required isl version 0.14 is deprecated
checking for isl 0.16 or 0.15... no
*** This configuration is not supported in the following subdirectories:
     target-libstdc++-v3 gnattools gotools target-libada target-libgfortran target-libbacktrace target-libgo target-libffi target-zlib target-libjava target-libobjc target-liboffloadmic target-boehm-gc target-libcilkrts target-libitm target-libsanitizer target-libvtv
    (Any other directories should still work fine.)

```

**Comment by grahamwhaley** [comment]
I tested this with the Centos image `CentOS-7-x86_64-Minimal-1611.iso` installed under QEMU (under Fedora24).
After @jodh-intel updated the section above adding the 'no passwd for sudo' extras, the install went smoothly (note, it takes some time...), and I was able to launch a CC container.
Notes:
- Initially I did not have nested VM enabled on my qemu/Centos, so launching failed
- It looked like maybe the proxy was not started - but that was probably a side effect of the nested-VM - I'll go check that again
- After a reboot of the Centos VM, docker is not started by default. I'll open a separate Issue for that.

I'll go check some more finer details, but overall this:
lgtm

**Comment by grahamwhaley** [comment]
Just confirming my initial proxy issue was due to not having nested-VM enabled.
All good for me.

**Comment by jodh-intel** [comment]
Thanks @grahamwhaley. @amshinde / @GabyCT - could you try once more on CentOS and RHEL respectively so we can get this merged and PR #771 unblocked.

@GabyCT - on the topic of #771, please can you update the doc with the minimal setup I've listed in https://github.com/01org/cc-oci-runtime/pull/799#issuecomment-292164923 (you'll need to change the git clone to be the master repo of course)?

**Comment by GabyCT** [comment]
I try it to run it again with the password and now it is not asking the password however it is not installing correctly in RHEL, I will send the typescript

**Comment by GabyCT** [comment]
@jodh-intel typescript sent it

**Comment by jodh-intel** [comment]
Hi @GabyCT - it looks like your rpm database is now inconsistent: your latest log shows that `yum` isn't installing the required packages.

**Comment by amshinde** [comment]
@jodh-intel I had checked out latest version of your repo. I'll give this a try again.

**Comment by amshinde** [comment]
@jodh-intel Tried this out again. Worked perfectly this time.
lgtm

**Comment by gorozco1** [comment]
lgtm

File: installation/rhel-setup.sh [hunk 1]
```diff
@@ -82,23 +87,40 @@ gcc_setup
...
+enabled=1
+gpgcheck=1
+gpgkey=https://yum.dockerproject.org/gpg
+EOF
+sudo yum install -y \
+    docker-engine-1.12.1-1.el7.centos.x86_64 \
+    docker-engine-selinux-1.12.1-1.el7.centos.noarch
+
 # Install kernel and CC image
 sudo yum-config-manager --add-repo http://download.opensuse.org/repositories/home:/clearlinux:/preview:/clear-containers-2.1/RHEL_7/home:clearlinux:preview:clear-containers-2.1.repo
```
Comment (gorozco1) [inline installation/rhel-setup.sh]: we need a conditional here to add proper repo depending of rhel or centos

```
if [ "$os_distribution" = rhel ]
then
	sudo yum-config-manager --add-repo http://download.opensuse.org/repositories/home:/clearlinux:/preview:/clear-containers-2.1/RHEL_7/home:clearlinux:preview:clear-containers-2.1.repo
elif [ "$os_distribution" = centos ]
then
	sudo yum-config-manager --add-repo http://download.opensuse.org/repositories/home:/clearlinux:/preview:/clear-containers-2.1/CentOS_7/home:clearlinux:preview:clear-containers-2.1.repo
fi

```

Comment (jodh-intel) [inline installation/rhel-setup.sh]: @gorozco1 - good catch! Although it's effectively currently a NOP it is clearly wrong and could have been the source of bugs in the future.

Fixed.

**Comment by jodh-intel** [comment]
Branch updated. Changes:

- Removed debug (`set -x`).
- Added correct CC repo (thanks @gorozco1).
- Added a brief description at the top of `rhel-setup.sh`.

**Comment by chavafg** [comment]
qa-passed

**Comment by devimc** [comment]
@jodh-intel  does it work in rhel? if yes could someone corroborate that?

**Comment by jodh-intel** [comment]
Hi @devimc - it should, but I am not able to test this.

@GabyCT - could you comment on RHEL?

**Comment by GabyCT** [comment]
the last time that I tried it, I was having issues (qemu and some other things were not building correctly), errors like the ones that I sent, so I decided to install again RHEL but I still do not test it. Maybe someone else can try it in order to see if the errors is only in my environment

**Comment by devimc** [comment]
@GabyCT Thanks
@jodh-intel lemme check in my environment

**Comment by gorozco1** [comment]
Thanks @jodh-intel lgtm

**Comment by devimc** [comment]
@jodh-intel 
```
checking for UUID... yes
checking for LIBMNL... no
configure: error: Package requirements (libmnl) were not met:

No package 'libmnl' found

Consider adjusting the PKG_CONFIG_PATH environment variable if you
installed software in a non-standard prefix.

Alternatively, you may set the environment variables LIBMNL_CFLAGS
and LIBMNL_LIBS to avoid the need to call pkg-config.
See the pkg-config man page for more details.
make: *** No targets specified and no makefile found.  Stop.
~/go/src/github.com/01org/cc-oci-runtime
```

**Comment by jodh-intel** [comment]
Hi @devimc - are you testing on CentOS or RHEL (I'm guessing RHEL)?

`rhel-setup.sh` installs `libmnl` and `libmnl-devel` from the archive. This works fine on CentOS, so could it be that those packages are not available under RHEL? Can you check the full `typescript` log to see if you got an error from `yum`?

File: installation/rhel-setup.sh [hunk 2]
```diff
@@ -58,6 +60,30 @@ pkgs+=" gcc-c++"
...
+    #
+    # Therefore, the safest approach seems to be to check if a known
+    # required development package is known to yum(8). If it isn't, the
+    # user will need to enable the extra repository.
+    #
+    # Note that this issue is unique to RHEL: yum on CentOS provides
+    # access to developemnt packages by default.
+
+    yum search -C "${mnl_dev_pkg}" >/dev/null 2>&1
+    if [ $? -ne 0 ]
```
Comment (devimc) [inline installation/rhel-setup.sh]: yum always returns 0 even if package does not exist

Comment (devimc) [inline installation/rhel-setup.sh]: yep, 'info' fixed the problem :)

**Comment by devimc** [comment]
lgtm

**Comment by chavafg** [comment]
qa-failed

**Comment by chavafg** [comment]
qa-passed

**Comment by dlespiau** [comment]
Kicked travis and it seems the certificate on gnome.org has been fixed. Given people have managed to run the script on CentOS and REHL, it seems like a good time to merge this, it's a vast improvement already.

**Pull request merged**